src/pd/etcd.rs

```diff
@@ -0,0 +1,355 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::error;
+use std::result;
+use std::io::{Read, Write, Error as IoError};
+use std::time::Duration;
+use std::collections::BTreeMap;
+
+use hyper::client::{Client, Handler, Request, Response, HttpConnector};
+use hyper::{Decoder, Encoder, Next, Error as HyperError, Post, StatusCode};
+use hyper::net::Transport;
+use hyper::header::ContentLength;
+use hyper::Url;
+use bytes::{Buf, MutBuf, ByteBuf, MutByteBuf};
+use rustc_serialize::json::{Json, ToJson};
+use rustc_serialize::base64::{FromBase64, ToBase64, STANDARD};
+use protobuf::Message;
+
+use kvproto::pdpb::Leader;
+
+use util::event::Event;
+use util::buf::{TryRead, TryWrite, create_mem_buf};
+use util::escape;
+
+const HTTP_REQUEST_TIMEOUT: u64 = 3;
+const ETCD_V3_KV_RANGE_PATH: &'static str = "/v3alpha/kv/range";
+
+quick_error!{
+    #[derive(Debug)]
+    pub enum Error {
+        Io(err: IoError) {
+            from()
+            cause(err)
+            description(err.description())
+        }
+        Hyper(err: HyperError) {
+            from()
+            cause(err)
+            description(err.description())
+        }
+        Status(status: StatusCode) {
+            description("invalid HTTP status code")
+            display("invalid HTTP status code {:?}", status)
+        }
+        Timeout(d: Duration) {
+            description("request timeout")
+            display("timeout after {:?}", d)
+        }
+        Other(err: Box<error::Error + Sync + Send>) {
+            from()
+            cause(err.as_ref())
+            description(err.description())
+            display("unknown error {:?}", err)
+        }
+    }
+}
+
+pub type Result<T> = result::Result<T, Error>;
+
+fn next_read() -> Next {
+    Next::read().timeout(Duration::from_secs(HTTP_REQUEST_TIMEOUT))
+}
+
+fn next_write() -> Next {
+    Next::write().timeout(Duration::from_secs(HTTP_REQUEST_TIMEOUT))
+}
+
+// Encode the key with base64 format and then covert
+// to Json object.
+pub fn b64encode_to_json(key: &[u8]) -> Json {
+    key.to_base64(STANDARD).to_json()
+}
+
+fn parse_response<T: Read>(r: &mut T) -> Result<Json> {
```

Comment:
I think use `try!(parse_response)` and `box_try!(Json::from_reader)` are the same. So no need to wrap another function.